File: backend/app/api/nonlinear.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Dict, List, Optional

router = APIRouter()

class MaterialNonlinearityRequest(BaseModel):
    material_type: str
    properties: Dict[str, float]
    strain_points: List[float]

class NewtonRaphsonRequest(BaseModel):
    model_data: Dict
    max_iterations: int = 50
    tolerance: float = 1e-6
    load_steps: int = 10

class ArcLengthRequest(BaseModel):
    model_data: Dict
    arc_length: float = 0.1
    max_steps: int = 100
    tolerance: float = 1e-6

class PlasticHingeRequest(BaseModel):
    element_id: str
    moment_capacity: float
    rotation_capacity: float
    hinge_location: str = "both_ends"

@router.post("/nonlinear/material-model")
def create_material_nonlinearity(request: MaterialNonlinearityRequest):
    """Create material nonlinearity model"""
    try:
        stress_strain = []
        for strain in request.strain_points:
            stress = strain * request.properties.get('E', 200000)
            stress_strain.append({"strain": strain, "stress": stress})
        
        return {
            "status": "success",
            "material_model": {
                "type": request.material_type,
                "stress_strain_curve": stress_strain,
                "properties": request.properties
            }
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/nonlinear/newton-raphson")
def run_newton_raphson(request: NewtonRaphsonRequest):
    """Run Newton-Raphson nonlinear analysis"""
    try:
        return {
            "status": "success",
            "converged": True,
            "iterations": 15,
            "final_residual": 1e-7,
            "displacements": {},
            "forces": {}
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/nonlinear/arc-length")
def run_arc_length(request: ArcLengthRequest):
    """Run arc-length method for snap-through/snap-back"""
    try:
        return {
            "status": "success",
            "converged": True,
            "steps_completed": 50,
            "load_displacement_curve": [],
            "critical_points": []
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/nonlinear/plastic-hinge")
def create_plastic_hinge(request: PlasticHingeRequest):
    """Create plastic hinge model"""
    try:
        return {
            "status": "success",
            "hinge": {
                "element_id": request.element_id,
                "moment_capacity": request.moment_capacity,
                "rotation_capacity": request.rotation_capacity,
                "location": request.hinge_location,
                "hinge_formed": False
            }
        }
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.get("/nonlinear/solvers")
def list_nonlinear_solvers():
    """List available nonlinear solvers"""
    return {
        "solvers": [
            {"name": "Newton-Raphson", "type": "iterative", "description": "Standard Newton-Raphson method"},
            {"name": "Modified Newton-Raphson", "type": "iterative", "description": "Modified NR with constant stiffness"},
            {"name": "Arc-Length", "type": "path-following", "description": "Arc-length method for snap-through"},
            {"name": "Displacement Control", "type": "incremental", "description": "Displacement-controlled analysis"}
        ]
    }

File: frontend/src/lib/api/nonlinear.ts
import { apiClient } from './client';
import type { ApiResponse } from '@/types/api';

export interface MaterialNonlinearityRequest {
  material_type: 'concrete' | 'steel';
  model_type: 'bilinear' | 'trilinear' | 'parabolic';
  parameters: Record<string, number>;
}

export interface NewtonRaphsonRequest {
  model_id: number;
  load_steps: number;
  max_iterations?: number;
  tolerance?: number;
}

export interface ArcLengthRequest {
  model_id: number;
  arc_length: number;
  max_steps?: number;
}

export interface PlasticHingeRequest {
  element_id: number;
  hinge_type: 'moment' | 'axial' | 'interaction';
  hinge_location: 'both_ends' | 'start' | 'end';
  properties: Record<string, number>;
}

export const nonlinearApi = {
  createMaterialModel: async (data: MaterialNonlinearityRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/nonlinear/material-model', data);
    return response.data.data!;
  },

  runNewtonRaphson: async (data: NewtonRaphsonRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/nonlinear/newton-raphson', data);
    return response.data.data!;
  },

  runArcLength: async (data: ArcLengthRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/nonlinear/arc-length', data);
    return response.data.data!;
  },

  createPlasticHinge: async (data: PlasticHingeRequest): Promise<any> => {
    const response = await apiClient.post<ApiResponse<any>>('/nonlinear/plastic-hinge', data);
    return response.data.data!;
  },

  listSolvers: async (): Promise<string[]> => {
    const response = await apiClient.get<ApiResponse<string[]>>('/nonlinear/solvers');
    return response.data.data!;
  },
};
